flink-runtime/src/main/java/org/apache/flink/runtime/state/CachedCheckpointStreamFactory.java

```diff
@@ -0,0 +1,195 @@
+		this.cache = cache;
+		this.remoteFactory = Preconditions.checkNotNull(factory, "Remote stream factory is null.");
+	}
+
+	public CheckpointStateOutputStream createCheckpointStateOutputStream(long checkpointID, long timestamp, StateHandleID handleID) throws Exception {
+		return createCheckpointStateOutputStream(checkpointID, timestamp, handleID, false);
+	}
+
+	public CheckpointStateOutputStream createCheckpointStateOutputStream(long checkpointID, long timestamp, StateHandleID handleID, boolean placeholder) throws Exception {
+		if (LOG.isDebugEnabled()) {
+			LOG.debug("create cache output stream: cpkID:{} placeHolder:{}", checkpointID, placeholder);
+		}
+		CachedOutputStream cachedOut = null;
+		if (cache != null) {
+			cachedOut = cache.createOutputStream(checkpointID, handleID, placeholder);
+		}
+		CheckpointStateOutputStream remoteOut = null;
+		if (!placeholder) {
+			remoteOut = remoteFactory.createCheckpointStateOutputStream(checkpointID, timestamp);
+		}
+		CachedCheckpointStateOutputStream output = new CachedCheckpointStateOutputStream(cachedOut, remoteOut);
+		return output;
+	}
+
+	@Override
+	public CheckpointStateOutputStream createCheckpointStateOutputStream(long checkpointID, long timestamp) throws Exception {
+		LOG.warn("create output stream which is not cacheable.");
+		return remoteFactory.createCheckpointStateOutputStream(checkpointID, timestamp);
+	}
+
+	@Override
+	public void close() throws Exception {
+		remoteFactory.close();
+	}
+
+	/**
+	 * A hybrid checkpoint output stream which write data to both remote end and local end,
+	 * writing data locally failed won't stop writing to remote. This hybrid output stream
+	 * will return a {@link CachedStreamStateHandle} in closeAndGetHandle(), it can be used for read data locally.
+	 */
+	public static class CachedCheckpointStateOutputStream extends CheckpointStateOutputStream {
+
+		private CachedOutputStream cacheOut = null;
+		private CheckpointStateOutputStream remoteOut = null;
+
+		public CachedCheckpointStateOutputStream(CachedOutputStream cacheOut, CheckpointStateOutputStream remoteOut) {
+			this.cacheOut = cacheOut;
+			this.remoteOut = remoteOut;
+		}
+
+		@Override
+		public StreamStateHandle closeAndGetHandle() throws IOException {
+			if (cacheOut != null) {
+				// finalize cache data
+				StateHandleID cacheId = cacheOut.getCacheID();
+				cacheOut.end();
+
+				StreamStateHandle remoteHandle;
+				if (remoteOut != null) {
+					remoteHandle = remoteOut.closeAndGetHandle();
+				} else {
+					remoteHandle = new PlaceholderStreamStateHandle(cacheId);
+				}
+				return new CachedStreamStateHandle(cacheId, remoteHandle);
+			} else {
+				if (remoteOut != null) {
+					return remoteOut.closeAndGetHandle();
+				} else {
+					return null;
+				}
+			}
+		}
+
+		@Override
+		public long getPos() throws IOException {
+			return remoteOut != null ? remoteOut.getPos() :-1L;
```

Comment:
This can be a bit dangerous. Some code may rely on certain offsets in the file, so both streams `cachedOut` and `remoteOut` better should have aligned offsets before we start. This should at least be checked in a precondition because it could lead to subtile, hard to debug bugs.

Reply:
What I means is that any seeks to offsets that we obtain by `getPos()` could fail on the cached stream if the offsets are not aligned.

Reply:
Thanks for point out this, addressing it.